torchinductor/ir.py

```diff
@@ -1477,20 +1477,24 @@ def simplify_reorder_and_tile(self):
                 var_ranges,
             )
         index_formulas = [*body.indexing_exprs.values()]
-        memory_addrs = [*body.reads_name2expr.values(), *body.writes_name2expr.values()]
+        # memory_addrs = [*body.reads_name2expr.values(), *body.writes_name2expr.values()]
+        # prioritize reads layout/loop_ordering over writes
+        if len(body.reads_name2expr.values()) > 0:
+            memory_addrs = [*body.reads_name2expr.values()]
```

Comment:
The motivation for using reads address only is to enabling potential fusion.
If we read from a NHWC layout, and write to NCHW layout. To make it fusible with the previous NHWC layout operation, we want to reorder loops to be the order of NHWC. But if we decide reordering based on both reads and writes, it is possible that it choose NCHW ordering. But this will prevent the possible fusion and create redundant read.
My heuristic here may not be good enough. Do you have any suggestions on reordering heuristics that could cover the case above while considering the general case?

Reply:
Have you benchmarked this to see if it slows anything down?

We could put it behind the `if config.triton.convolution != "aten"` flag if needed.  

It might not matter since we aren't doing layout tuning yet.

Reply:
But it shouldn't prevent fusions based on just layout? That's what tiling is for, so that you can efficiently read NHWC inputs and produce NCHW output (or vice versa)

Reply:
for output size as arg0 [N, H, W, C] = [s0, 55, 55, 64]
Assume buf0 = arg0.to(memory_layout=torch.contigous); buf1 = buf0.clone(); They will be fused into 1 kernel
```
x0 = xindex % 64          
x1 = (xindex // 64) % 3025  
x2 = xindex // 193600 
x5 = xindex % 193600 (equivalent to x5 = x0 + x1 *64)
x6 = xindex // 193600
tmp = tl.load(in_ptr0 + xindex) # read from conv arg0 N(HW)C layout
tl.store(out_ptr0 + x1 + 3025 * x0 + 193600 * x2, tmp) # write to buf0 NC(HW) layout
tl.store(out_ptr1 + x5 + 193600 * x6, tmp) # want to write to buf1 NC(HW) layout, but wrongly write to NHWC layout
```
This is because the indexing x5 = xindex % 193600 is generated unware of loop order of buf0. To buf1's view, it reads from buf0 in NCHW layout and write to buf1 NCHW layout, and generate the x5 and x6 indexing code. But because of cse.store_cache[arg0] = tmp, buf1 use` tmp` value, but still use the x5, x6 indexing, which is wrong.
Using the index_prevent_reordering when generating `MemoryDep`, it will break into 2 kernels. But it we decide loop reordering to be channel-last for buf1. Then they still could be fused into a single kernel.
Since it will prevent the fusion if we do not reorder the loops wisely, I prioritize reads layout/loop order in the heuristics. But this won't affect default NCHW layout case right now.

Reply:
I'm sorry, I still don't understand. Introducing the concept of layouts when we already have the concept of strides is very confusing. With strides, tensor sizes are NCHW always, the strides correspond to either channels first or channels last layout. That way, iteration domains are always aligned, and we never should fuse ops with the different iteration domains, and correct accesses are generated using the strides.
I still don't understand how code above would be generated, because for 

```
def convert_fn(x0):
    buf0 = x0.contiguous() #to(memory_format=torch.contiguous_format) - errors in lowering
    buf1 = buf0.clone()
    return buf0, buf1

x=torch.rand(4, 64, 55, 55, device=device, dtype=dtype).to(memory_format=torch.channels_last)
with optimize_ctx:
    convert_fn(x)
```
I'm getting the code producing correct values (albeit non-contiguous outputs, despite me specifically requesting contiguous outputs). But I think the confusion stems from trying to operate on tensors with the different sizes - NCHW and NHWC, and this is inherently unsafe and will be very fragile.

Reply:
In your snippet above 
```
buf0 = arg0.to(memory_layout=torch.contigous)
```
if arg0 sizes are (s0, 55, 55, 64) as you write, buf0 will also have sizes (s0, 55, 55, 64), and both buf0 and arg0 would be interpreted in pytorch as tensors with 55 channels and 55x64 images. You might be looking at them differently only in you codepaths, but then all the differences should be localized to generated convolution codepaths, and not spill out.

Reply:
Sorry the code I show above is not representative. For the code
```
cnn = nn.Conv2d(3, 64, kernel_size=11, stride=4, padding=0)
def cnn_concat(x1, x2, x3):
    y1 = cnn(x1) # buf0 = y1
    y2 = cnn(x2) # buf1 = y2
    y3 = cnn(x3) # buf2 = y3

    return (torch.cat([y1, y2], dim=1), torch.cat([y1, y3], dim=1))
x1 = torch.randn([32, 3, 224, 224], device="cuda", dtype=torch.float32)
x2 = torch.randn([32, 3, 224, 224], device="cuda", dtype=torch.float32)
x3 = torch.randn([32, 3, 224, 224], device="cuda", dtype=torch.float32)
y_correct = run(x1, x2, x3)
torchdynamo.reset()
torchinductor.config.triton.convolution = "triton"
with torch.no_grad():
    y = run_inductor(x1, x2, x3)
assert(same(y, y_correct, cos_similarity=True))
```
On main branch, it will codegen kernel4 as 
```
@pointwise_heuristics(size_hints=[131072, 64])
@triton.jit
def kernel4(in_ptr0, in_ptr1, in_ptr2, in_ptr3, out_ptr0, out_ptr1, out_ptr2, out_ptr3, xnumel, ynumel, XBLOCK : tl.constexpr, YBLOCK : tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.reshape(tl.arange(0, XBLOCK), [XBLOCK, 1])
    xmask = xindex < xnumel
    yoffset = tl.program_id(1) * YBLOCK
    yindex = yoffset + tl.reshape(tl.arange(0, YBLOCK), [1, YBLOCK])
    ymask = yindex < ynumel
    x3 = xindex
    y2 = yindex
    x0 = xindex % 3025
    x1 = (xindex // 3025)
    tmp0 = tl.load(in_ptr0 + y2 + (64*x3) + tl.zeros([XBLOCK, YBLOCK], tl.int32), xmask & ymask)
    tmp1 = tl.load(in_ptr1 + y2 + tl.zeros([XBLOCK, YBLOCK], tl.int32), xmask & ymask)
    tmp3 = tl.load(in_ptr2 + y2 + (64*x3) + tl.zeros([XBLOCK, YBLOCK], tl.int32), xmask & ymask)
    tmp5 = tl.load(in_ptr3 + y2 + (64*x3) + tl.zeros([XBLOCK, YBLOCK], tl.int32), xmask & ymask)
    tmp2 = tmp0 + tmp1
    tmp4 = tmp3 + tmp1
    tmp6 = tmp5 + tmp1
    tl.store(out_ptr0 + x0 + (3025*y2) + (387200*x1) + tl.zeros([XBLOCK, YBLOCK], tl.int32), tmp2, xmask & ymask)
    tl.store(out_ptr1 + x0 + (3025*y2) + (387200*x1) + tl.zeros([XBLOCK, YBLOCK], tl.int32), tmp4, xmask & ymask)
    tl.store(out_ptr2 + x0 + (3025*y2) + (387200*x1) + tl.zeros([XBLOCK, YBLOCK], tl.int32), tmp6, xmask & ymask)
    tl.store(out_ptr3 + y2 + (64*x0) + (387200*x1) + tl.zeros([XBLOCK, YBLOCK], tl.int32), tmp6, xmask & ymask)
```

Switched to this PR, it will codegen
```
@pointwise_heuristics(size_hints=[131072, 64])
@triton.jit
def kernel4(in_ptr0, in_ptr1, in_ptr2, in_ptr3, out_ptr0, out_ptr1, out_ptr2, out_ptr3, ks0, xnumel, ynumel, XBLOCK : tl.constexpr, YBLOCK : tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.reshape(tl.arange(0, XBLOCK), [XBLOCK, 1])
    xmask = xindex < xnumel
    yoffset = tl.program_id(1) * YBLOCK
    yindex = yoffset + tl.reshape(tl.arange(0, YBLOCK), [1, YBLOCK])
    ymask = yindex < ynumel
    x3 = xindex
    y2 = yindex
    x0 = xindex % 3025
    x0_next = xindex // 3025
    x1 = x0_next
    tmp0 = tl.load(in_ptr0 + y2 + (64*x3) + tl.zeros([XBLOCK, YBLOCK], tl.int32), xmask & ymask)
    tmp1 = tl.load(in_ptr1 + y2 + tl.zeros([XBLOCK, YBLOCK], tl.int32), xmask & ymask)
    tmp3 = tl.load(in_ptr2 + y2 + (64*x3) + tl.zeros([XBLOCK, YBLOCK], tl.int32), xmask & ymask)
    tmp5 = tl.load(in_ptr3 + y2 + (64*x3) + tl.zeros([XBLOCK, YBLOCK], tl.int32), xmask & ymask)
    tmp2 = tmp0 + tmp1
    tmp4 = tmp3 + tmp1
    tmp6 = tmp5 + tmp1
    tl.store(out_ptr0 + x0 + (3025*y2) + (387200*x1) + tl.zeros([XBLOCK, YBLOCK], tl.int32), tmp2, xmask & ymask)
    tl.store(out_ptr1 + x0 + (3025*y2) + (387200*x1) + tl.zeros([XBLOCK, YBLOCK], tl.int32), tmp4, xmask & ymask)
    tl.store(out_ptr2 + x0 + (3025*y2) + (387200*x1) + tl.zeros([XBLOCK, YBLOCK], tl.int32), tmp6, xmask & ymask)
    tl.store(out_ptr3 + x0 + (3025*y2) + (387200*x1) + tl.zeros([XBLOCK, YBLOCK], tl.int32), tmp2, xmask & ymask)
```

The only difference is the last store indexing expression. When changing the reordering, we basically change the indexing forumula